### PR TITLE
Text editor: enable UTF-8 search

### PR DESCRIPTION
**Pull request by hius07**
Continue and thanks to https://github.com/koreader/koreader/pull/7966#issuecomment-880831477

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7981)

### CODE REVIEW

**Comment by hius07** [comment]
No crashes with the test file.

**Comment by hius07** [comment]
Manual input is safe, but we can copy a broken char from the text, paste it into the search input and crash the program.
Maybe, add the check for the search string too?

**Comment by poire-z** [comment]
Looks fine to me.

Btw, another (minor) issue that you may not be able to solve as you're not (yet) emulator aware:
when I type with my physical keyboard (the emulator supports that for normal keys), what I enter in the search field also goes into the main text :)
image

I must say I'm a little puzzled looking at the code - I have no idea how/where my keystrokes are handled and end up into some InputText :/

**Comment by poire-z** [comment]
> Maybe, add the check for the search string too?

Yes, why not ? :)

File: frontend/ui/widget/inputtext.lua [hunk 1]
```diff
@@ -673,19 +673,19 @@ end
...
-        str = Utf8Proc.lowercase(str)
-    end
     for i = start_pos, #self.charlist - str_len do
         for j = 1, str_len do
             local char_txt = self.charlist[i + j]
+            local char_str = str_charlist[j]
             if not case_sensitive then
-                char_txt = Utf8Proc.lowercase(char_txt)
+                char_txt = Utf8Proc.lowercase(util.fixUtf8(char_txt,"?"))
+                char_str = Utf8Proc.lowercase(util.fixUtf8(char_str,"?"))
```
Comment (NiLuJe) [inline frontend/ui/widget/inputtext.lua:686]: Style nit: space after comma, please ;).

Comment (hius07) [reply frontend/ui/widget/inputtext.lua:686]: Sure

**Comment by poire-z** [comment]
>  I have no idea how/where my keystrokes are handled and end up into some InputText :/

Ok, found it:
https://github.com/koreader/koreader/blob/9d2c6b07aa4e93bf4d25b583384e700bd315e145/frontend/ui/widget/inputtext.lua#L563-L568

and we have a broadcast that won't be stopped with `return true`:
https://github.com/koreader/koreader/blob/9d2c6b07aa4e93bf4d25b583384e700bd315e145/frontend/device/sdl/device.lua#L266-L268

**Comment by poire-z** [comment]
And it's solved with just:
```diff
--- a/frontend/device/sdl/device.lua
+++ b/frontend/device/sdl/device.lua
@@ -264,7 +264,7 @@ function Device:init()
                 self.window.left = ev.value.data1
                 self.window.top = ev.value.data2
             elseif ev.code == SDL_TEXTINPUT then
-                UIManager:broadcastEvent(Event:new("TextInput", ev.value))
+                UIManager:sendEvent(Event:new("TextInput", ev.value))
             end
         end,
         hasClipboardText = function()
```
@NiLuJe : is it really as simple as that ? :)
(There is not a single explicite `UIManager:sendEvent()` currently, only UIManager uses `self:sendEvent()`...)

**Comment by NiLuJe** [comment]
@poire-z: That'll stop it as the first handler that returns true ;).

(.e.g, like most widgets that do it the other way around with a `self:handleEvent`, except with send you don't know the receiver in advance: it'll be the uppermost widget with a matching handler)

**Pull request merged**